File: node_js_ds_and_algos_implementations/stacks/generate_parenthesis.js
class Solution {
    /**
     * @param {number} n
     * @return {string[]}
     */
    generateParenthesis(n) {
        /*
        given n,
        you need n opening parenthesis and n closing parenthesis present
        to be considered a valid string of parenthesis

        if you add an opening
        (
         in the next spot, you can add either another opening or closing 

        if there are more opening braces than closing braces (and there are still closing braces available), 
        you can add a closing brace

        example execution:
        n = 3
        ((())) is the first one that's found (since we continually pick opening braces, then closing braces)
        (( - in this case, we have 2 openings and 0 closings, so the amount of openings > amount of closings, so we can add a closing brace
        (() - now at this step, we have 2 openings and 1 closing, we have the choice of adding an opening brace, or a closing brace,

      (()( - adds opening, this is valid -> branches into another recursion tree
      (()) - adds closing, this is also valid -> branches into another recursion tree
        */

        let res = []
        const search = (numOpen, numClose, cur) => {
            if (numOpen === n && numClose === n){
                res.push(cur.join(""))
                return
            }
            // always try adding an opening brace first
            if (numOpen < n){
                search(numOpen+1, numClose, [...cur, "("])
            }
            // add closing if the amount of opening braces is greater than the amount of closing
            // and there are still closing braces
            if (numOpen > numClose && numClose < n){
                search(numOpen, numClose+1, [...cur, ")"])
            }
        }
        search(0,0, [])
        return res

    }
}

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    /*
    using backtracking
    given you have n pairs of parenthesis, you will have n*2 total,
    with n opening braces, and n closing braces
    
    decision tree
    you always start by using an opening brace
    (
    
    at this point, you can either:
    1) add another opening brace, subtract one from total num of opening braces
    2) add a closing brace, subtract one from total num of closing braces
    
        (
     ()   ((
     
     Note at with the left side, we can't add another closing brace since we've 
     created a valid pair so far, so we can only add an opening brace
     
        (
     ()   ((
     
   ()(  (() (((
    
    Note on the right most side (((, we've run out of opening braces, so we must pick closing braces only
    
    at a given recursive call, we can tell whether we have 2 options (add opening or closing), or 1 option (adding only opening),
    if we keep a track of a stack. Whenever we add an opening brace, we add to the stack. If we add a closing brace,
    instead of adding to the top of the stack, we pop off the stack. 
    
    Exponential Time Complexity
    */
    let res = []
    const helper = (cur, numLeft, numRight, stack) => {
        if (numLeft <= 0 && numRight <= 0){
            res.push(cur)
            return
        }
        if (stack.length > 0){   
            if (stack[stack.length-1] === "("){
                if (numLeft > 0){
                    helper(cur + "(", numLeft - 1, numRight, stack.concat("("))
                }
                if (numRight > 0){
                    helper(cur + ")", numLeft, numRight - 1, stack.splice(1, stack.length-1))
                }
            }
        }
        else {
            if (numLeft > 0){
                helper(cur + "(", numLeft - 1, numRight, stack.concat("("))
            }
        }
        
    }
    // we always start with an opening brace no matter what, otherwise the entire string would be considered invalid
    helper("(", n-1, n, ["("])
    return res
};

File: neetcode all and misc questions/Dynamic Programming and Recursion/generate_parenthesis.py
'''
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

Example 1:

Input: n = 3
Output: ["((()))","(()())","(())()","()(())","()()()"]
Example 2:

Input: n = 1
Output: ["()"]

'''
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
        Backtracking:
        n = 3
        ((())) first case
        backtracks to ...
        (( openingCount = 2 and closingCount = 3
        as long as there are more closing than opening,
        we can choose closing as an option here. 
        (()
        this next call, we can go back to choosing an opening brace by default
        since opening braces always come before closing
        """
        res = []
        def search(openingCount: int, closingCount: int, cur: str) -> None:
            # base case: no more braces to choose here
            if openingCount == 0 and closingCount == 0:
                res.append(cur)
                return
            # default to choose opening brace before closing
            if openingCount > 0:
                search(openingCount - 1, closingCount, cur + "(")
            # if we're out of opening braces and there are 
            # still closing braces, choose closing brace
            # also if we have more closing braces than opening,
            # we can also choose a closing brace
            if (openingCount == 0 and closingCount > 0) or (closingCount > openingCount):
                search(openingCount, closingCount - 1, cur + ")")

        search(n,n,"")
        return res

"""
Revisited on 9/26/2024, still same solution in 2^n time
"""
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def search(leftCount, rightCount, cur, res):
            if (leftCount == n and rightCount == n):
                res.append(cur)
                return
            if (leftCount < n):
                search(leftCount+1, rightCount, cur + "(", res)
            """ 
            as long as there are more left parens then right parens, we can use a right parens,
            this prevents us from choosing a right parens before a left parens, which would
            create an invalid combination
            """
            if (leftCount > rightCount and rightCount < n):
                search(leftCount, rightCount+1, cur + ")", res)
        valid = []
        search(0, 0, "", valid)
        return valid
"""
Revisited on 9/8/2023
Time: O(2^n), for a given index there can be two combinations here, so 2 * 2 * 2 ... n which is 2^n
Space: O(2^n)
1) The concept is that we start with one opening brace, and then 
we can continually pick an opening brace until the number of opening braces is equal to our n
2) Then, we can continually pick closing braces
3) To make sure we don't accidentally pick a closing brace and then an opening brace,
we check that the amount of opening braces is greater than our amount of closing braces,
so we ensure that we always pick opening braces before closing braces in our recursion
4) If our number of openings and number of closings are both equal to n, we have found formed a
combination with all the pairs, so save into the global array
5) return the global array
"""
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []
        def dfs(numOpening, numClosing, curr):
            # if the number of openings == n and number of closing parens == n,
            # save current combination to our result and stop
            if numOpening == n and numClosing == n:
                self.res.append(curr)
                return
            # if we can continue picking opening brackets, keep picking opening brackets
            if numOpening < n:
                dfs(numOpening + 1, numClosing, curr + "(")
            """
            if we have more opening braces than closing braces and we can
            pick closing braces still, pick the closing brace. This is to prevent us 
            from picking a closing brace and then an opening brace if there are an
            equal amount of opening and closing braces, thus forming
            an invalid combination 
            Example: (i.e  "())(" )
            if n = 2, after creating (())
            the recursion backtracks to (, and then adds a closing brace to form (),
            which then creates ()(). 
            After backtracking on this case, it'd get to the single pair of braces (),
            where numOpening == numClosing.
            The numOpening > numClosing condition prevents us from choosing a closing brace
            here like so: ())
            """
            if numOpening > numClosing and numClosing < n:
                dfs(numOpening, numClosing + 1, curr + ")")
        # we always start with one opening brace, so numOpening starts at 1
        dfs(1, 0, "(")
        return self.res

class Solution:
    ## Concept: brute force, generate every possible combination of parenthesis, but only accept
    ## the ones that create a valid parenthesis structure
    ## Time complexity: O(2^2n * n), generate 2^2n sequences, the extra 2n since there's two characters for the set of parens, multiplied by an additional N to check for a valid parenthesis
    ## Space complexity: O(2^2n * n) to hold every valid sequence
    def __init__(self):
        self.combinations = []
    def generateParenthesis(self, n: int) -> List[str]:
        self.generate(float(n), '')
        return list(self.combinations)
    def generate(self, n: float, cur: str):
        if (n == 0):
            if (self.isValidParenthesis(cur)):
                self.combinations.append(cur)
        else:
            self.generate(n-.5, cur + '(')
            self.generate(n-.5, cur + ')')
                
            
    def isValidParenthesis(self,cur):
        i = 0
        j = len(cur)
        s = []
        for i in range(len(cur)): 
            if (len(s) > 0):
                if (cur[i] == ')' and s[-1] == '('):
                    s.pop()
                    continue
            s.append(cur[i])
        return len(s) == 0
        